SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -785,10 +843,109 @@ void USpatialReceiver::ApplyComponentData(Worker_EntityId EntityId, Worker_Compo
 	TWeakObjectPtr<UObject> TargetObject = PackageMap->GetObjectFromUnrealObjectRef(FUnrealObjectRef(EntityId, Offset));
 	if (!TargetObject.IsValid())
 	{
-		UE_LOG(LogSpatialReceiver, Warning, TEXT("EntityId %lld, ComponentId %d, Offset %d - Could not find target object with given offset for Actor %s!"), EntityId, Data.component_id, Offset, *Channel->GetActor()->GetName());
+		// If we can't find this subobject, it's a dynamically attached object. Create it now.
+		TargetObject = NewObject<UObject>(Channel->GetActor(), ClassInfoManager->GetClassByComponentId(Data.component_id));
+
+		Channel->GetActor()->OnSubobjectCreatedFromReplication(TargetObject.Get());
+
+		PackageMap->ResolveSubobject(TargetObject.Get(), FUnrealObjectRef(EntityId, Offset));
+
+		Channel->CreateSubObjects.Add(TargetObject.Get());
+	}
+
+	ApplyComponentData(TargetObject.Get(), Channel, Data);
+}
+
+void USpatialReceiver::HandleDynamicAddComponent(Worker_AddComponentOp& Op)
+{
+	// If we are delegated a component that exists, we will receive an AddComponentOp.
+	// This will be a duplicate of the AddComponentOp we've already recieved in the AddEntity payload
+	// So it can be safely ignored.
+	if (StaticComponentView->HasComponent(Op.entity_id, Op.data.component_id))
+	{
+		return;
+	}
+
+	uint32 Offset = 0;
+	bool bFoundOffset = ClassInfoManager->GetOffsetByComponentId(Op.data.component_id, Offset);
+	if (!bFoundOffset)
+	{
+		UE_LOG(LogSpatialReceiver, Warning, TEXT("EntityId %lld, ComponentId %d - Could not find offset for component id when receiving dynamic AddComponent."), Op.entity_id, Op.data.component_id);
```

Comment:
Again, even without strict line limits I think we shouldn't use more than about 180 characters to avoid having to scroll right on our monitors.

Reply:
ZOOM OUT